libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp

```diff
@@ -273,12 +276,55 @@ void NavEKF3_core::SelectMagFusion()
             } else {
                 fuseEulerYaw(true, true);
             }
+            have_fused_gps_yaw = true;
+            last_gps_yaw_fusion_ms = now;
         }
-        return;
+        if (magcal == MagCal::EXTERNAL_YAW) {
+            // no fallback
+            return;
+        }
+
+        // get new mag data into delay buffer
+        readMagData();
+
+        if (have_fused_gps_yaw) {
+            if (gps_yaw_mag_fallback_active) {
+                gps_yaw_mag_fallback_active = false;
+                gcs().send_text(MAV_SEVERITY_INFO, "EKF3 IMU%u yaw external",(unsigned)imu_index);
+            }
+            // update mag bias from GPS yaw
+            gps_yaw_mag_fallback_ok = learnMagBiasFromGPS();
+            return;
+        }
+
+        // we don't have GPS yaw data and are configured for
+        // fallback. If we've only just lost GPS yaw
+        if (now - last_gps_yaw_fusion_ms < 5000) {
```

Comment:
I think we can increase this to at least 10 seconds to give the GPS more time to recover. Yaw drift due to uncompensated bias will be less than 0.5 deg/sec in most situations

Reply:
changed to 10s